bindings/python/scripts/parse.py

```diff
@@ -1,145 +1,89 @@
-            if child_list and parsed_list:
-                if len(parsed_list[0]["members"]):
-                    parsed_list[0]["members"].append(child_list[0])
-                else:
-                    parsed_list[0]["members"] = child_list
+        if node_in_this_file(node=child, filename=this_filename):
+            child_parsed_info = generate_parsed_info(cursor=child, this_filename=this_filename, depth=depth + 1,)
+            if child_parsed_info and parsed_info:
+                parsed_info["members"].append(child_parsed_info)
 
+    return parsed_info
 
-def parse_arguments(args):
-    parser = argparse.ArgumentParser(description="C++ libclang parser")
-    parser.add_argument("files", nargs="+", help="The source files to search")
-    return parser.parse_args(args)
 
+def create_index():
+    return clang.Index.create()
 
-def get_output_path(source, output_dir):
-    x_list = source.split("pcl/", 1)[-1]
-    x_list = x_list.split("/")
-    extra = ["pcl", "include"]
 
-    filename = x_list[-1].split(".")[0]
-    relative_dir = "/".join(x for x in x_list[:-1] if x not in extra)
-    dir = os.path.join(output_dir, relative_dir)
+def get_compilation_database(compile_db_path):
+    return clang.CompilationDatabase.fromDirectory(buildDir=compile_db_path)
 
-    # ensure the new directory exists
-    if not os.path.exists(dir):
-        os.makedirs(dir)
 
-    return f"{dir}/{filename}.json"
+def get_compile_commands(compdb, filename):
+    compile_commands = compdb.getCompileCommands(filename=filename)
+    # extracting argument list from the command's generator object
+    return list(compile_commands[0].arguments)[1:-1]
 
 
 def main():
-    args = parse_arguments(sys.argv[1:])
-    index = clang.Index.create()
+    args = utils.parse_arguments(args=sys.argv[1:], script=__file__)
+    for source in args.files:
+        source = utils.get_realpath(path=source)
 
-    compile_db_path = os.path.dirname("./compile_commands.json")
-    compdb = clang.CompilationDatabase.fromDirectory(compile_db_path)
+        index = create_index()
 
-    for source in args.files:
-        with open(source) as input_file:
-            lines = input_file.readlines()
-
-        compile_commands = compdb.getCompileCommands(source)
-        # extracting argument list from the command's generator object
-        compile_commands = list(compile_commands[0].arguments)[1:-1]
-        tu = index.parse(source, args=compile_commands)
-
-        # walk_and_print(
-        #     tu.cursor, filter, lines, tu.spelling, depth=0
-        # )
-
-        parsed_list = []
-        generate_parsed_info(
-            tu.cursor, filter, lines, tu.spelling, depth=0, parsed_list=parsed_list,
-        )
-
-        output_filepath = get_output_path(
-            os.path.realpath(source), output_dir=f"json/{os.path.dirname(__file__)}"
-        )
-        dump_json(output_filepath, parsed_list)
+        compile_db_path = utils.get_dirname(path="../compile_commands.json")
```

Comment:
Good point. They are while running the code from different directories.
Something like 1ec5411 should work.

Reply:
Eh, now the code only works on your machine though, since e.g. CI runs somewhere that doesn't have `/home/divyanshu/Projects/active` prepended to the `/pcl`-bit

Reply:
A suggestion would be appreciated :slightly_smiling_face:

Reply:
Use `os.path` or something similar in Python. I'd recommend allowing the user to provide the script, and having a sane default

Reply:
I let the user specify the path to `compile_commands.json` and the directory where they want to store the output JSON(s), the defaults for both being the parent directory of the `scripts`.


10b7c55 and 50a3a06

Reply:
I googled "relative path python", [this is the first result](https://stackoverflow.com/questions/918154/relative-paths-in-python) and it seems good enough to me. Do notice some of the comments on portability, as CI runs on Windows too.

Reply:
I've used the same as the link :slightly_smiling_face: